// @eth-optimism/viem/dist/core/baseWriteAction.js
import { estimateContractGas, writeContract } from 'viem/actions';
export async function baseWriteAction(client, contractParameters, parameters) {
    const { abi, contractAddress, contractFunctionName, contractArgs } = contractParameters;
    const { account, chain = client.chain, gas, maxFeePerGas, maxPriorityFeePerGas, nonce, value, accessList, } = parameters;
    const gas_ = typeof gas !== 'bigint' && gas !== null
        ? await estimateContractGas(client, {
            account: account,
            abi,
            address: contractAddress,
            chain,
            functionName: contractFunctionName,
            args: contractArgs,
            maxFeePerGas,
            maxPriorityFeePerGas,
            nonce,
            value,
            accessList,
        })
        : gas ?? undefined;
    return writeContract(client, {
        account: account,
        abi,
        address: contractAddress,
        chain,
        functionName: contractFunctionName,
        args: contractArgs,
        gas: gas_,
        maxFeePerGas,
        maxPriorityFeePerGas,
        nonce,
        value,
        accessList,
    });
}
