// @eth-optimism/viem/dist/addressSet.js
export const addressesToViemContractConstant = (addressSet, sourceId) => {
    return {
        opChainProxyAdmin: addressForChain(addressSet.OpChainProxyAdmin, sourceId),
        addressManager: addressForChain(addressSet.AddressManager, sourceId),
        l1Erc721BridgeProxy: addressForChain(addressSet.L1ERC721BridgeProxy, sourceId),
        systemConfig: addressForChain(addressSet.SystemConfigProxy, sourceId),
        optimismMintableErc20FactoryProxy: addressForChain(addressSet.OptimismMintableERC20FactoryProxy, sourceId),
        l1StandardBridge: addressForChain(addressSet.L1StandardBridgeProxy, sourceId),
        l1CrossDomainMessenger: addressForChain(addressSet.L1CrossDomainMessengerProxy, sourceId),
        optimismPortal: addressForChain(addressSet.OptimismPortalProxy, sourceId),
        disputeGameFactory: addressForChain(addressSet.DisputeGameFactoryProxy, sourceId),
        anchorStateRegistry: addressForChain(addressSet.AnchorStateRegistryProxy, sourceId),
        faultDisputeGame: addressForChain(addressSet.FaultDisputeGame, sourceId),
        permissionedDisputeGame: addressForChain(addressSet.PermissionedDisputeGame, sourceId),
    };
};
const addressForChain = (address, chainId) => {
    return {
        [chainId]: { address },
    };
};
